### ISSUE TITLE
Provide Automatic-Module-Name attribute in MANIFEST.MF

### ISSUE THREAD

**Issue by afloarea** [vert-x3/issues#524]
#### Feature Description

Provide Automatic-Module-Name attribute in MANIFEST.MF file.

#### Use cases

Provides a reliable module name, instead of using the name derived from the JAR file, without having to target Java 9+ and/or use module-info java/class file. This can be used until there is a module-info file.

#### Contribution

If this feature request were to be approved I would be happy to work on it.

I believe that this change would require some additional configuration in the pom.xml maven jar plugin, for example:

```

<plugin>
    <groupId>org.apache.maven.plugins</groupId>
    <artifactId>maven-jar-plugin</artifactId>
    <configuration>
        <archive>
            <manifestEntries>
                <Automatic-Module-Name>io.vertx.core</Automatic-Module-Name>
            </manifestEntries>
        </archive>
    </configuration>
</plugin>

```

#### Naming Scheme

When choosing module names, [these rules](https://sormuras.github.io/blog/2018-11-16-invalid-automatic-module-names) should be taken into consideration.

Table of automatic module names and their status:

Notes:
 - The deprecated MySQL / PostgreSQL clients are not included.
 - AMPQ Bridge is not included as it will be removed in Vertx 4.
 - Module names cannot contain `-`, `..` or Java keywords, [as this article suggests](https://sormuras.github.io/blog/2018-11-16-invalid-automatic-module-names#how-to-name-a-module).

## Core
Component | Artifact id | Automatic Module Name | Status
--- | --- | --- | --- 
Vertx core | vertx-core | io.vertx.core | [Merged](https://github.com/eclipse-vertx/vert.x/pull/3401)

## Web
Component | Artifact id | Automatic Module Name | Status 
--- | --- | --- | --- 
Vertx web | vertx-web | io.vertx.web | [Merged](https://github.com/vert-x3/vertx-web/pull/1608)
Vertx web client | vertx-web-client | io.vertx.web.client | [Merged](https://github.com/vert-x3/vertx-web/pull/1608)
Vertx web common | vertx-web-common | io.vertx.web.common | [Merged](https://github.com/vert-x3/vertx-web/pull/1608)
Vertx web OpenAPI | vertx-web-openapi | io.vertx.web.openapi | [Merged](https://github.com/vert-x3/vertx-web/pull/1516)
Vertx web validation | vertx-web-validation | io.vertx.web.validation | [Merged](https://github.com/vert-x3/vertx-web/pull/1608)
Vertx web templates | vertx-web-templ-[x] | io.vertx.web.template.[x] | [Merged](https://github.com/vert-x3/vertx-web/pull/1608)
Vertx web API | vertx-web-api-contract | io.vertx.web.apicontract | [Merged](https://github.com/vert-x3/vertx-web/pull/1608)
Vertx web API service | vertx-web-api-service | io.vertx.web.apiservice | [Merged](https://github.com/vert-x3/vertx-web/pull/1608)
Web GraphQL Handler | vertx-web-graphql | io.vertx.web.graphql | [Merged](https://github.com/vert-x3/vertx-web/pull/1608)
Web sstores | vertx-web-sstore-[x] | io.vertx.web.sstore.[x] | [Merged](https://github.com/vert-x3/vertx-web/pull/1608)
Web Proxy | vertx-web-proxy | io.vertx.web.proxy | Todo
Web OpenAPI Rtouer | vertx-web-openapi-router | io.vertx.web.openapi.router | Todo

## Data access

Component | Artifact id | Automatic Module Name | Status 
--- | --- | --- | --- 
MongoDB client | vertx-mongo-client | io.vertx.client.mongo | [Merged](https://github.com/vert-x3/vertx-mongo-client/pull/228)
Redis client | vertx-redis-client | io.vertx.client.redis | [Merged](https://github.com/vert-x3/vertx-redis-client/pull/231)
Cassandra client | vertx-cassandra-client | io.vertx.client.cassandra | [Merged](https://github.com/vert-x3/vertx-cassandra-client/pull/57)
Common SQL interfaces | vertx-sql-common | N/A | [Merged](https://github.com/vert-x3/vertx-sql-common/issues/69)
JDBC client | vertx-jdbc-client | io.vertx.client.jdbc | [Merged](https://github.com/vert-x3/vertx-jdbc-client/pull/184)

### Reactive SQL clients
Component | Artifact id | Automatic Module Name | Status 
--- | --- | --- | --- 
SQL client | vertx-sql-client | io.vertx.client.sql | [Merged](https://github.com/eclipse-vertx/vertx-sql-client/pull/641)
DB2 client | vertx-db2-client | io.vertx.client.sql.db2 | [Merged](https://github.com/eclipse-vertx/vertx-sql-client/pull/641)
MSSQL client | vertx-mssql-client | io.vertx.client.sql.mssql | [Merged](https://github.com/eclipse-vertx/vertx-sql-client/pull/641)
MySQL client | vertx-mysql-client | io.vertx.client.sql.mysql | [Merged](https://github.com/eclipse-vertx/vertx-sql-client/pull/641)
PostgreSQL client | vertx-pg-client | io.vertx.client.sql.pg | [Merged](https://github.com/eclipse-vertx/vertx-sql-client/pull/641)
Templates | vertx-client-sql-templates | io.vertx.client.sql.templates | [Merged](https://github.com/eclipse-vertx/vertx-sql-client/pull/641)

## Reactive
Component | Artifact id | Automatic Module Name | Status 
--- | --- | --- | --- 
Reactive Streams | vertx-reactive-streams | io.vertx.reactivestreams | [Merged](https://github.com/vert-x3/vertx-reactive-streams/pull/19)
Vertx Sync | vertx-sync | N/A | - 
Kotlin coroutines | vertx-lang-kotlin-coroutines | io.vertx.kotlin.coroutines | [Merged](https://github.com/vert-x3/vertx-lang-kotlin/pull/169)

### Reactive Extensions

Component | Artifact id | Automatic Module Name | Status 
--- | --- | --- | --- 
Rx code generator | vertx-rx-gen | io.vertx.rx.gen | Postponed
RxJava code generator | vertx-rx-java-gen | io.vertx.rx.java.gen | Postponed
RxJava API | vertx-rx-java | io.vertx.rx.java | Postponed 
RxJava2 Code Generator | vertx-rx-java2-gen | io.vertx.rx.java2.gen | Postponed
RxJava2 API | vertx-rx-java2 | io.vertx.rx.java2 | Postponed
JUnit 5 RxJava support | vertx-junit5-rx-java | io.vertx.junit5.rx.java | Postponed
JUnit 5 RxJava2 support | vertx-junit5-rx-java2 | io.vertx.junit5.rx.java2 | Postponed

## Microservices
Component | Artifact id | Automatic Module Name | Status 
--- | --- | --- | --- 
Circuit Breaker | vertx-circuit-breaker | io.vertx.circuitbreaker | [Merged](https://github.com/vert-x3/vertx-circuit-breaker/pull/45)

### Vertx Config
Component | Artifact id | Automatic Module Name | Status 
--- | --- | --- | --- 
Vertx Config | vertx-config | io.vertx.config | [Merged](https://github.com/vert-x3/vertx-config/pull/118)
Consul | vertx-config-consul | io.vertx.config.consul | [Merged](https://github.com/vert-x3/vertx-config/pull/118)
Git | vertx-config-git | io.vertx.config.git | [Merged](https://github.com/vert-x3/vertx-config/pull/118)
HOCON | vertx-config-hocon | io.vertx.config.hocon | [Merged](https://github.com/vert-x3/vertx-config/pull/118)
Kubernetes ConfigMap | vertx-config-kubernetes-configmap | io.vertx.config.configmap | [Merged](https://github.com/vert-x3/vertx-config/pull/118)
Redis | vertx-config-redis | io.vertx.config.redis | [Merged](https://github.com/vert-x3/vertx-config/pull/118)
Spring config server | vertx-config-spring-config-server | io.vertx.config.springconfigserver | [Merged](https://github.com/vert-x3/vertx-config/pull/118)
Vault | vertx-config-vault | io.vertx.config.vault | [Merged](https://github.com/vert-x3/vertx-config/pull/118)
YAML | vertx-config-yaml | io.vertx.config.yaml | [Merged](https://github.com/vert-x3/vertx-config/pull/118)
Zookeeper | vertx-config-zookeeper | io.vertx.config.zookeeper | [Merged](https://github.com/vert-x3/vertx-config/pull/118)

### Vertx Service Discovery
Component | Artifact id | Automatic Module Name | Status 
--- | --- | --- | --- 
Service Discovery | vertx-service-discovery | io.vertx.servicediscovery | [Merged](https://github.com/vert-x3/vertx-service-discovery/pull/124/files)
Consul Backend | vertx-service-discovery-backend-consul | io.vertx.servicediscovery.backend.consul | [Merged](https://github.com/vert-x3/vertx-service-discovery/pull/124)
Redis Backend | vertx-service-discovery-backend-redis | io.vertx.servicediscovery.backend.redis | [Merged](https://github.com/vert-x3/vertx-service-discovery/pull/124)
Zookeeper Backend | vertx-service-discovery-backend-zookeeper | io.vertx.servicediscovery.backend.zookeeper | [Merged](https://github.com/vert-x3/vertx-service-discovery/pull/124)
Consul Bridge | vertx-service-discovery-bridge-consul | io.vertx.servicediscovery.bridge.consul | [Merged](https://github.com/vert-x3/vertx-service-discovery/pull/124)
Docker links Bridge | vertx-service-discovery-bridge-docker-links | io.vertx.servicediscovery.bridge.dockerlinks | [Merged](https://github.com/vert-x3/vertx-service-discovery/pull/124)
Docker Bridge | vertx-service-discovery-bridge-docker | io.vertx.servicediscovery.bridge.docker | [Merged](https://github.com/vert-x3/vertx-service-discovery/pull/124)
Kubernetes Bridge | vertx-service-discovery-bridge-kubernetes | io.vertx.servicediscovery.bridge.kubernetes | [Merged](https://github.com/vert-x3/vertx-service-discovery/pull/124)
Zookeeper Bridge | vertx-service-discovery-bridge-zookeeper | io.vertx.servicediscovery.bridge.zookeeper | [Merged](https://github.com/vert-x3/vertx-service-discovery/pull/124)

## Standards

Component | Artifact id | Automatic Module Name | Status 
--- | --- | --- | --- 
Vertx OpenAPI | vertx-openapi | io.vertx.openapi | Done
Vertx Json Schema | vertx-json-schema | io.vertx.jsonschema | Todo
Vertx URI Template | vertx-uri-temlate | io.vertx.uritemplate | Todo

## HTTTP Proxy

Component | Artifact id | Automatic Module Name | Status 
--- | --- | --- | --- 
Vertx Http Proxy | vertx-http-proxy| io.vertx.httpproxy | Todo

## MQTT
Component | Artifact id | Automatic Module Name | Status 
--- | --- | --- | --- 
Vertx MQTT | vertx-mqtt | io.vertx.mqtt | [Merged](https://github.com/vert-x3/vertx-mqtt/pull/157)

## Authentication and Authorisation
Component | Artifact id | Automatic Module Name | Status 
--- | --- | --- | --- 
Auth Common | vertx-auth-common | io.vertx.auth.common | [Merged](https://github.com/vert-x3/vertx-auth/pull/395)
JDBC Auth | vertx-auth-jdbc | io.vertx.auth.jdbc | [Merged](https://github.com/vert-x3/vertx-auth/pull/395)
JWT Auth | vertx-auth-jwt | io.vertx.auth.jwt | [Merged](https://github.com/vert-x3/vertx-auth/pull/395)
JWT | vertx-jwt | N/A | [Artifact code merged into vertx-auth-common](https://github.com/vert-x3/vertx-auth/issues/388)
LDAP | vertx-auth-ldap | io.vertx.auth.ldap | [Merged](https://github.com/vert-x3/vertx-auth/pull/395)
Shiro Auth | vertx-auth-shiro | io.vertx.auth.shiro | [Merged](https://github.com/vert-x3/vertx-auth/pull/395)
MongoDB Auth | vertx-auth-mongo | io.vertx.auth.mongo | [Merged](https://github.com/vert-x3/vertx-auth/pull/395)
OAuth 2 | vertx-auth-oauth2 | Io.vertx.auth.oauth2 | [Merged](https://github.com/vert-x3/vertx-auth/pull/395)
Properties | vertx-auth-properties | Io.vertx.auth.properties | [Merged](https://github.com/vert-x3/vertx-auth/pull/395)
.htdigest Auth | vertx-auth-htdigest | io.vertx.auth.htdigest | [Merged](https://github.com/vert-x3/vertx-auth/pull/395)
.htpasswd | vertx-auth-htpasswd | io.vertx.auth.htpasswd | [Merged](https://github.com/vert-x3/vertx-auth/pull/395)
SQL Auth | vertx-auth-sql | io.vertx.auth.sql | [Merged](https://github.com/vert-x3/vertx-auth/pull/395)
WebAuthN | vertx-auth-webauthn | io.vertx.auth.webauthn | [Merged](https://github.com/vert-x3/vertx-auth/pull/395)

## Messaging
Component | Artifact id | Automatic Module Name | Status 
--- | --- | --- | --- 
AMQP Client | vertx-amqp-client | io.vertx.client.amqp | [Merged](https://github.com/vert-x3/vertx-amqp-client/pull/34)
STOMP | vertx-stomp | io.vertx.stomp | [Merged](https://github.com/vert-x3/vertx-stomp/pull/67)
RabbitMQ Client | vertx-rabbitmq-client | io.vertx.client.rabbitmq | [Merged](https://github.com/vert-x3/vertx-rabbitmq-client/pull/104)
Vert.x Proton | vertx-proton | io.vertx.proton | [Merged](https://github.com/vert-x3/vertx-proton/pull/112)

## Integration

Component | Artifact id | Automatic Module Name | Status 
--- | --- | --- | --- 
Kafka Client | vertx-kafka-client | io.vertx.client.kafka | [Merged](https://github.com/vert-x3/vertx-kafka-client/pull/174)
Mail Client | vertx-mail-client | io.vertx.client.mail | [Merged](https://github.com/vert-x3/vertx-mail-client/pull/146) 
Consul Client | vertx-consul-client | io.vertx.client.consul | [Merged](https://github.com/vert-x3/vertx-consul-client/pull/77)

### JCA Adaptor
Component | Artifact id | Automatic Module Name | Status 
--- | --- | --- | --- 
JCA Adapter | vertx-jca-adapter | N/A | -
JCA API | vertx-jca-api | N/A | -

## Event Bus Bridge
Component | Artifact id | Automatic Module Name | Status 
--- | --- | --- | --- 
Common | vertx-bridge-common | io.vertx.eventbusbridge.common | [Merged](https://github.com/vert-x3/vertx-bridge-common/pull/6)
TCP Event Bus Bridge | vertx-tcp-eventbus-bridge | io.vertx.eventbusbridge.tcp | Under discussion
Camel Bridge | vertx-camel-bridge | io.vertx.eventbusbridge.camel | Under discussion

## Devops

Component | Artifact id | Automatic Module Name | Status 
--- | --- | --- | --- 
Dropwizard metrics | vertx-dropwizard-metrics | io.vertx.metrics.dropwizard | [Merged](https://github.com/vert-x3/vertx-dropwizard-metrics/pull/98)
Micrometer | vertx-micrometer-metrics | io.vertx.metrics.micrometer | [Merged](https://github.com/vert-x3/vertx-micrometer-metrics/pull/104)
Health Check | vertx-health-check | io.vertx.healthcheck | [Merged](https://github.com/vert-x3/vertx-health-check/pull/27)
Shell | vertx-shell | io.vertx.shell | [Merged](https://github.com/vert-x3/vertx-shell/pull/44)
Docker | vertx-stack-docker | N/A | - 
Stack Manager | vertx-stack-manager | N/A | -

## Tracing
Component | Artifact id | Automatic Module Name | Status 
--- | --- | --- | --- 
Zipkin | vertx-zipkin | io.vertx.tracing.zipkin | [Merged](https://github.com/eclipse-vertx/vertx-tracing/pull/5)
Opentracing | vertx-opentracing | io.vertx.tracing.opentracing | [Merged](https://github.com/eclipse-vertx/vertx-tracing/pull/5)

## Testing
Component | Artifact id | Automatic Module Name | Status 
--- | --- | --- | --- 
Vertx Unit | vertx-unit | io.vertx.testing.unit | [Merged](https://github.com/vert-x3/vertx-unit/pull/70)
JUnit 5 | vertx-junit5 | io.vertx.testing.junit5 | [Merged](https://github.com/vert-x3/vertx-junit5/pull/79)

## Clustering
Component | Artifact id | Automatic Module Name | Status 
--- | --- | --- | --- 
Hazelcast | vertx-hazelcast | io.vertx.clustermanager.hazelcast | [Merged](https://github.com/vert-x3/vertx-hazelcast/pull/128)
Infinispan | vertx-infinispan | io.vertx.clustermanager.infinispan | [Merged](https://github.com/vert-x3/vertx-infinispan/pull/76)
Apache Ignite | vertx-ignite | io.vertx.clustermanager.ignite | [Merged](https://github.com/vert-x3/vertx-ignite/pull/86)
Zookeeper | vertx-zookeeper | io.vertx.clustermanager.zookeeper | [Merged](https://github.com/vert-x3/vertx-zookeeper/pull/97)

## Services

Component | Artifact id | Automatic Module Name | Status 
--- | --- | --- | --- 
Service Proxies | vertx-service-proxy | io.vertx.serviceproxy | [Merged](https://github.com/vert-x3/vertx-service-proxy/pull/114)
Service Factories | vertx-service-factory | N/A | - 
HTTP Service Factory | vertx-http-service-factory | N/A | -
SockJs Service Proxy | vertx-sockjs-service-proxy | io.vertx.serviceproxy.sockjs | [Merged](https://github.com/vert-x3/vertx-sockjs-service-proxy/pull/11)
Maven Service Factory | vertx-maven-service-factory | N/A | - 

### gRPC
Component | Artifact id | Automatic Module Name | Status 
--- | --- | --- | --- 
gRPC | vertx-grpc | io.vertx.grpc | [Merged](https://github.com/vert-x3/vertx-grpc/pull/69)
gRPC | vertx-grpc-client | io.vertx.grpc.client | Done
gRPC | vertx-grpc-server | io.vertx.grpc.server | Done
gRPC | vertx-grpc-common | io.vertx.grpc.common | Done
gRPC | vertx-grpc-context-storage | N/A | -
gRPC protoc plugin | vertx-grpc-protoc-plugin | N/A | - 

## Internal Magic
Component | Artifact id | Automatic Module Name | Status 
--- | --- | --- | --- 
Codegen | vertx-codegen | io.vertx.codegen | [Merged](https://github.com/vert-x3/vertx-codegen/pull/302)
Docgen | vertx-docgen | N/A | - 
Codetrans | vertx-codetrans | N/A | -
Bridge Common | vertx-bridge-common | io.vertx.eventbusbridge.common | [Merged](https://github.com/vert-x3/vertx-bridge-common/pull/6) 
Distro Stack Manager | vertx-stack-manager | io.vertx.stack.manager | N/A

## Languages
Component | Artifact id | Automatic Module Name | Status 
--- | --- | --- | --- 
Groovy lang | vertx-lang-groovy | N/A | N/A

### Kotlin

Component | Artifact id | Automatic Module Name | Status 
--- | --- | --- | --- 
Kotlin gen | vertx-lang-kotlin-gen | io.vertx.kotlin.gen | Under discussion
Kotling lang | vertx-lang-kotlin | io.vertx.kotlin | [Merged](https://github.com/vert-x3/vertx-lang-kotlin/pull/169)

### Scala

Component | Artifact id | Automatic Module Name | Status 
--- | --- | --- | --- 
Scala lang codegen | vertx-lang-scala-codegen | N/A | -
Scala language support | vertx-lang-scala_2.13 | N/A | -

**Comment by vietj** [comment]
I believe we cannot use package names because some package have the ext name and some don't, which is inconsistent. So I would like module to be rather named after the related Maven artifact, likely prefixed with `io.vertx` instead of `vertx-`, e.g

- `vertx-core`: `io.vertx.core`
- `vertx-web`: `io.vertx.web`
- `vertx-web-client`: `io.vertx.web.client`

That being said, are we allowed to have a module named `io.vert.web` and another one named `io.vertx.web.client` ?

There is in the rules defines by https://sormuras.github.io/blog/2018-11-16-invalid-automatic-module-names#how-to-name-a-module a rule I do not understand: _Creating a module with a particular name takes ownership of that package name and everything beneath it._:

- what is the supposed relationship between module names and package names ? I think there is not such relationship, i.e one can create a module named `foo.bar` that contains a `com.whatever` package
- what means _everything beneath it_ ?

**Comment by afloarea** [comment]
I believe that there is no issue with having automatic modules/submodules, such as `io.vertx.web` and `io.vertx.web.client` as long as they do not have any split packages between them. 

A thing to note: Say you have 2 artifacts/jars/libraries A and B which have a split package P (they both have classes in package P).

If they do not have a module name (via MANIFEST property) and are placed on the module path, then they will get mashed up into the unnamed module. The client JPMS application will run. 

If however, they had module names (via MANIFEST), they would each become a (automatic) module. The client JPMS application will not run in this case because of the split package. This outcome would be the same if module A would be named but module B would not have a name, since they will end up in different modules again.

Client code can require/use both `io.vertx.web` and `io.vertx.web.client` as long as the 2 jars/artifacts/modules do not have any split packages.

In regards to the ownership of the package name and everything beneath it - What I understand from this is that if you choose the package name `io.vertx` then you are responsible for creating unique modules under that name (`io.vertx.x.y.z`) that do not have split packages.

There is no strict relationship between the module name and the package name. It is only a recommandation that facilitates module name uniqueness.

 Considering your example, you can indeed create a module `foo.bar` that contains a `com.whatever` package. Since you own whatever.com, no one else would create classes under that package , but they could use the module name `foo.bar`, say with com.something.else package. A client application would not be able however to use both modules at the same time, even thought there are totally different packages in the 2 modules and that is because the modules have the same name.

To summarize, I believe that the requirement is to have unique modules names under io.vertx that do not have split packages. Maybe more useful/detailed information could be found here: https://blog.joda.org/2017/04/java-se-9-jpms-module-naming.html

So yes, the artifact name prefixed with io.vertx is a viable option, but the problem of split packages must be considered.

**Comment by vietj** [comment]
So the rule _Creating a module with a particular name takes ownership of that package name and everything beneath it._ is not a JPMS rule to satisfy as I understand it ?

**Comment by afloarea** [comment]
No, I believe it is not. You are right.

**Comment by vietj** [comment]
ok, so let's get first a proposal for providing automatic module names to vertx jars (when it makes sense) and then we can see how this can be achieved in the stack which will require to update the vertx-parent pom

**Comment by afloarea** [comment]
I am not sure I understand. I think your suggestion is good, that is, the module name would be `io.vertx.${processed-artifact-name}`, where the processed artifact name would be the artifact name without the `vertx-` prefix and with `-` converted to `.` (i.e. the `vertx-jdbc-client-<version>` jar would have the module name `io.vertx.jdbc.client`)

**Comment by vietj** [comment]
by proposal I mean list all module names according to this convention and then see discuss this in community to see if there are issues with that

**Comment by afloarea** [comment]
Ok, well here is the list so far ( I will edit this comment to keep the list up to date based on your suggestions )

## Core
Component | Artifact id | Automatic Module Name
--- | --- | ---
Vertx core | vertx-core | io.vertx.core

## Web
Component | Artifact id | Automatic Module Name
--- | --- | ---
Vertx web | vertx-web | io.vertx.web
Vertx web client | vertx-web-client | io.vertx.web.client
Vertx web API | vertx-web-api-contract | io.vertx.web.api.contract
Vertx web API service | vertx-web-api-service | io.vertx.web.api.service
Web GraphQL Handler | vertx-web-graphql | io.vertx.web.graphql

## Data access

_Note_: I did not include the deprecated MySQL / PostgreSQL clients.

Component | Artifact id | Automatic Module Name
--- | --- | ---
MongoDB client | vertx-mongo-client | io.vertx.mongo.client
Redis client | vertx-redis-client | io.vertx.redis.client
Cassandra client | vertx-cassandra-client | io.vertx.cassandra.client
Common SQL interfaces | vertx-sql-common | io.vertx.sql.common
JDBC client | vertx-jdbc-client | io.vertx.jdbc.client

### Reactive SQL clients
Component | Artifact id | Automatic Module Name
--- | --- | ---
DB2 client | vertx-db2-client | io.vertx.db2.client
MSSQL client | vertx-mssql-client | io.vertx.mssql.client
MySQL client | vertx-mysql-client | io.vertx.mysql.client
PostgreSQL client | vertx-pg-client | io.vertx.pg.client
SQL client | vertx-sql-client | io.vertx.sql.client

## Reactive
Component | Artifact id | Automatic Module Name
--- | --- | ---
Reactive Streams | vertx-reactive-streams | io.vertx.reactive.streams
Vertx Sync | vertx-sync | io.vertx.sync
Kotlin coroutines | vertx-lang-kotlin-coroutines | io.vertx.lang.kotlin.coroutines

### Reactive Extensions

Component | Artifact id | Automatic Module Name
--- | --- | ---
Rx code generator | vertx-rx-gen | io.vertx.rx.gen
RxJava code generator | vertx-rx-java-gen | io.vertx.rx.java.gen
RxJava API | vertx-rx-java | io.vertx.rx.java
RxJava2 Code Generator | vertx-rx-java2-gen | io.vertx.rx.java2.gen
RxJava2 API | vertx-rx-java2 | io.vertx.rx.java2
JUnit 5 RxJava support | vertx-junit5-rx-java | io.vertx.junit5.rx.java
JUnit 5 RxJava2 support | vertx-junit5-rx-java2 | io.vertx.junit5.rx.java2

## Microservices
Component | Artifact id | Automatic Module Name
--- | --- | ---
Circuit Breaker | vertx-circuit-breaker | io.vertx.circuit.breaker

### Vertx Config
Component | Artifact id | Automatic Module Name
--- | --- | ---
Vertx Config | vertx-config | io.vertx.config
Consul | vertx-config-consul | io.vertx.config.consul
Git | vertx-config-git | io.vertx.config.git
HOCON | vertx-config-hocon | io.vertx.config.hocon
Kubernetes ConfigMap | vertx-config-kubernetes-configmap | io.vertx.config.kubernetes.configmap
Redis | vertx-config-redis | io.vertx.config.redis
Spring config server | vertx-config-spring-config-server | io.vertx.config.spring.config.server
Vault | vertx-config-vault | io.vertx.config.vault
YAML | vertx-config-yaml | io.vertx.config.yaml
Zookeeper | vertx-config-zookeeper | io.vertx.config.zookeeper

### Vertx Service Discovery
Component | Artifact id | Automatic Module Name
--- | --- | ---
Service Discovery | vertx-service-discovery | io.vertx.service.discovery
Consul Backend | vertx-service-discovery-backend-consul | io.vertx.service.discovery.backend.consul
Redis Backend | vertx-service-discovery-backend-redis | io.vertx.service.discovery.backend.redis
Zookeeper Backend | vertx-service-discovery-backend-zookeeper | io.vertx.service.discovery.backend.zookeeper
Consul Bridge | vertx-service-discovery-bridge-consul | io.vertx.service.discovery.bridge.consul
Docker links Bridge | vertx-service-discovery-bridge-docker-links | io.vertx.service.discovery.bridge.docker.links
Docker Bridge | vertx-service-discovery-bridge-docker | io.vertx.service.discovery.bridge.docker
Kubernetes Bridge | vertx-service-discovery-bridge-kubernetes | io.vertx.service.discovery.bridge.kubernetes
Zookeeper Bridge | vertx-service-discovery-bridge-zookeeper | io.vertx.service.discovery.bridge.zookeeper

## MQTT
Component | Artifact id | Automatic Module Name
--- | --- | ---
Vertx MQTT | vertx-mqtt | io.vertx.mqtt

## Authentication and Authorisation
Component | Artifact id | Automatic Module Name
--- | --- | ---
Auth Common | vertx-auth-common | io.vertx.auth.common
JDBC Auth | vertx-auth-jdbc | io.vertx.auth.jdbc
JWT Auth | vertx-auth-jwt | io.vertx.auth.jwt
Shiro Auth | vertx-auth-shiro | io.vertx.auth.shiro
MongoDB Auth | vertx-auth-mongo | io.vertx.auth.mongo
OAuth 2 | vertx-auth-oauth2 | Io.vertx.auth.oauth2
.htdigest Auth | vertx-auth-htdigest | io.vertx.auth.htdigest

## Messaging

_Note_: I did not include AMPQ Bridge as it will be removed in Vertx 4.

Component | Artifact id | Automatic Module Name
--- | --- | ---
AMQP Client | vertx-amqp-client | io.vertx.amqp.client
STOMP | vertx-stomp | io.vertx.stomp
RabbitMQ Client | vertx-rabbitmq-client | io.vertx.rabbitmq.client

## Integration

_Note_: The link to the Vertx Mail Client repo from the [documentation page](https://vertx.io/docs/) is broken.

Component | Artifact id | Automatic Module Name
--- | --- | ---
Kafka Client | vertx-kafka-client | io.vertx.kafka.client
Mail Client | vertx-mail-client | io.vertx.mail.client
Consul Client | vertx-consul-client | io.vertx.consul.client

### JCA Adaptor
Component | Artifact id | Automatic Module Name
--- | --- | ---
JCA Adapter | vertx-jca-adapter | io.vertx.jca.adapter
JCA API | vertx-jca-api | io.vertx.jca.api

## Event Bus Bridge
Component | Artifact id | Automatic Module Name
--- | --- | ---
TCP Event Bus Bridge | vertx-tcp-eventbus-bridge | io.vertx.tcp.eventbus.bridge
Camel Bridge | vertx-camel-bridge | io.vertx.camel.bridge

## Devops

_Note_: The link to the Vertx Health Check repo from the [documentation page](https://vertx.io/docs/) is broken.

Component | Artifact id | Automatic Module Name
--- | --- | ---
Dropwizard metrics | vertx-dropwizard-metrics | io.vertx.dropwizard.metrics
Micrometer | vertx-micrometer-metrics | io.vertx.micrometer.metrics
Health Check | vertx-health-check | io.vertx.health.check
Shell | vertx-shell | io.vertx.shell
Docker | vertx-stack-docker | io.vertx.stack.docker
Stack Manager | vertx-stack-manager | io.vertx.stack.manager

## Testing
Component | Artifact id | Automatic Module Name
--- | --- | ---
Vertx Unit | vertx-unit | io.vertx.unit
JUnit 5 | vertx-junit5 | io.vertx.junit5

## Clustering
Component | Artifact id | Automatic Module Name
--- | --- | ---
Hazelcast | vertx-hazelcast | io.vertx.hazelcast
Infinispan | vertx-infinispan | io.vertx.infinispan
Apache Ignite | vertx-ignite | io.vertx.ignite
Zookeeper | vertx-zookeeper | io.vertx.zookeeper

## Services

_Note_: The link to the SockJS Service Proxies repo from the [documentation page](https://vertx.io/docs/) is broken.

Component | Artifact id | Automatic Module Name
--- | --- | ---
Service Proxies | vertx-service-proxy | io.vertx.service.proxy
Service Factories | vertx-service-factory | io.vertx.service.factory
HTTP Service Factory | vertx-http-service-factory | io.vertx.http.service.factory
SockJs Service Proxy | vertx-sockjs-service-proxy | io.vertx.sockjs.service.proxy
Maven Service Factory | vertx-maven-service-factory | io.vertx.maven.service.factory

### gRPC
Component | Artifact id | Automatic Module Name
--- | --- | ---
gRPC | vertx-grpc | io.vertx.grpc
gRPC protoc plugin | vertx-grpc-protoc-plugin | io.vertx.grpc.protoc.plugin

## Internal Magic
Component | Artifact id | Automatic Module Name
--- | --- | ---
Codegen | vertx-codegen | io.vertx.codegen
Docgen | vertx-docgen | io.vertx.docgen
Codetrans | vertx-codetrans | io.vertx.codetrans
Bridge Common | vertx-bridge-common | io.vertx.bridge.common
Distro Stack Manager | vertx-stack-manager | io.vertx.stack.manager

## Languages
Component | Artifact id | Automatic Module Name
--- | --- | ---
Groovy lang | vertx-lang-groovy | io.vertx.lang.groovy
JS lang gen | vertx-lang-js-gen | io.vertx.lang.js.gen

### Ruby
Component | Artifact id | Automatic Module Name
--- | --- | ---
Ruby lang gen | vertx-lang-ruby-gen | io.vertx.lang.ruby.gen
Ruby MVEL | vertx-lang-ruby-mvel | io.vertx.lang.ruby,mvel

### Kotlin

_Note_: There are 2 repositories with kotlin coroutines?

Component | Artifact id | Automatic Module Name
--- | --- | ---
Kotlin gen | vertx-lang-kotlin-gen | io.vertx.lang.kotlin.gen
Kotling lang | vertx-lang-kotlin | io.vertx.lang.kotlin

### Scala

Component | Artifact id | Automatic Module Name
--- | --- | ---
Scala lang codegen | vertx-lang-scala-codegen | io.vertx.lang.scala.codegen
Scala language support | vertx-lang-scala_2.13 | *__io.vertx.lang.scala_2.13__*

## Unclassified
Component | Artifact id | Automatic Module Name
--- | --- | ---
Vert.x Proton | vertx-proton | io.vertx.proton

I might have added too many. Hopefully I did not miss any. You have a LOT of stuff :)

**Comment by vietj** [comment]
I think there are a few names we shall override, e.g

- `io.vertx.reactive.streams` -> `io.vertx.reactive_streams`
- `io.vertx.sql.client` -> `io.vertx.sqlclient` or `io.vertx.sql_client`
- `io.vertx.dropwizard.metrics` -> `io.vertx.dropwizard_metrics`
- `io.vertx.micrometer.metrics` -> `io.vertx.micrometer_metrics` or `io.vertx.micrometer`

etc...

**Comment by afloarea** [comment]
I have updated the reactive streams entry.

Which of the 2 sql client options would be preferred?

In my opinion, it would be helpful to keep the metrics module names consistent. That is, `io.vertx.micrometer` & `io.vertx.dropwizard` or `io.vertx.micrometer_metrics` & `io.vertx.dropwizard_metrics`. Again, which of these would be preferred?

Also, what should be the work flow? Should work start when the entire list is green lit, or start work on approved modules names so that work on the approved module names and the approval of other module names happen in parallel?

I am just not sure of the next steps. Can you please provide some insight information?

**Comment by vietj** [comment]
I think we can do both in parallel.

For the packaging, can we study the modification of the `vertx-parent` to have the automatic module name applied optionally ? i.e if we define a property in the pom.xml then there is an automatic module otherwise it is left outside (I believe there are some module that won't benefit from automatic module name such as vertx-lang-groovy).

**Comment by vietj** [comment]
Perhaps for keeping track of this we should use the wiki, is that possible to get PR on a GitHub wiki ?

**Comment by afloarea** [comment]
Thank you for the prompt response @vietj . I will look into the change to the `vertx-parent` and raise a PR on the GitHub wiki and get back to you.

**Comment by vietj** [comment]
please do an RFC like page in wiki

**Comment by afloarea** [comment]
I have created the RFC page, but I cannot push it to a remote branch in order to create PR because I do not have permission. I assumed that [this](https://github.com/vert-x3/wiki/wiki) is the wiki.

I have also looked into changing the parent pom so that when a property in the child pom is defined, then the module name is applied. Unfortunately, I could not find a way for this to work, since maven profiles cannot be activated by a property defined in a pom.xml, only via system property, file, etc, so the closest thing would be `mvn package -Dautomatic.module=io.vertx.core`.

In my opinion, it would be easier / more straightforward to just add the `<automatic.module>io.vertx.whatever</automatic.module>` as a maven property in the child pom and also add the jar plugin configuration in the child pom for the default-jar execution and not modify the parent pom at all.

**Comment by vietj** [comment]
@afloarea where are your changes ?

**Comment by afloarea** [comment]
I have cloned the wiki and created a local branch on my machine but I cannot push the branch because I do not have permission to push to the wiki repo. 
Is there a different way to contribute that I am not aware of?

**Comment by vietj** [comment]
let me investigate how the wiki thing work :-)

**Comment by vietj** [comment]
what if a problem should not have an automatic name ? e.g we have some jars which will never be modules.

**Comment by vietj** [comment]
@afloarea unfortunately the wiki access management of GitHub is quite weak, so the best I think is that you edit this issue content instead of commenting and we continue using this issue to discuss

**Comment by afloarea** [comment]
> what if a problem should not have an automatic name ? e.g we have some jars which will never be modules.

I agree, not every jar file needs an automatic module name.
My proposal is that based on the component/project table, a name is approved and marked as approved in the table. After that, a PR is made for only that component that will edit the pom.xml file of that project/component only. The PR would simply add an `<automatic.module.name>` property in the `<properties>` section of the pom.xml and add the additional jar-plugin configuration of the manifest and use that property. The PR would only affect the child pom.xml of the component, so only one component.

So, for example, say you approve the io.vertx.core name, then a PR would be made to the child pom.xml of the vertx-core component which will add the automatic module name. Upon merge, only the vertx-core jar would have an automatic module name.

If, for example, it is decided that the vertx-lang-groovy jar should remain without an automatic module name, then no PR is raised for that component's pom.xml file, and the table is simply updated with 'not applicable' or something similar and that's that. It will remain as is (i.e. without a name)

What do you think?

**Comment by vietj** [comment]
+1 Adrian

**Comment by vietj** [comment]
can you edit this topic and move here the table and add checks ?

**Comment by afloarea** [comment]
Of course. You are referring to the table from above, right? When you say 'add checks' do you want me to add checkboxes or do you want me to mark some entries as checked / confirmed name.

I was thinking that you could confirm some of the names and I would update the table and with your approval I would start working on a PR for the confirmed names and possibly provide a link to the PR in the table.

**Comment by vietj** [comment]
@afloarea I mean in the first message of the issue you created
by check box I mean using this:

* [ ] foo
* [ ] bar

**Comment by afloarea** [comment]
@vietj I have moved the table to the top and I have added check boxes as you suggested.

**Comment by afloarea** [comment]
Hey @vietj . I have went ahead and raised a PR for adding automatic module name for vertx-core -> https://github.com/eclipse-vertx/vert.x/pull/3401

Could you please confirm other module names so that I may start working on those as well? Is it ok to carry on with the web category?

Also, would it be possible to shorten the contribution flow? I saw that for small changes it may be possible to contribute directly to master. It's a bit cumbersome to fork the project, create a branch with one commit with 2 additional lines and then raise a PR.

**Comment by vietj** [comment]
@afloarea for the name I think we need to do a meeting with the team and discuss that, e.g I feel that for clients we should maybe name to `io.vertx.client.mongo` rather than `io.vertx.mongo.client`

**Comment by vietj** [comment]
@afloarea I've setup an hangout meeting next Monday to agree on the name of modules, if you give me your email you can join the meeting

**Comment by afloarea** [comment]
@vietj Thank you for the invite. My email is adrianfloarea07@yahoo.com

**Comment by vietj** [comment]
you should receive an invite.

**Comment by vietj** [comment]
See https://github.com/vert-x3/vertx-auth/issues/388

**Comment by vietj** [comment]
See https://github.com/vert-x3/vertx-sql-common/issues/69

**Comment by vietj** [comment]
@afloarea vertx 4 projects have been updated with new parent pom

**Comment by afloarea** [comment]
Thanks @vietj . I have updated the PR for vertx-core. I will proceed in a similar fashion with the web artifacts.

**Comment by vietj** [comment]
@afloarea I updated the tables with more approvals

**Comment by vietj** [comment]
can you point out the pending PR to merge ?

**Comment by afloarea** [comment]
@vietj The PR should appear on this issue. They are just above your comment where you say that you updated the tables.

Currently vertx-mongo-client has been merged, the redis, cassandra and jdbc clients have an open PR and the web artifactis also have an open PR, but I guess that is on hold until the openAPI gets merged.

Also, I have seen that you approved more service discovery artifacts. I forgot to update the suggested automatic module name for those that you approved and they were left with sevice.discovery in the module name instead of servicediscovery (without the dot). I will change them to the latter (just FYI, let me know if you did actually want them with the dot and I will change them back)

I will also update the table with any open PRs.

**Comment by vietj** [comment]
@afloarea I don't see any pending PR linked in this document, am I missing something ?

**Comment by afloarea** [comment]
@vietj 

![image](https://user-images.githubusercontent.com/20104805/82139256-3c410380-982f-11ea-855b-6240144abf29.png)

I am also updating the table as we speak.

**Comment by afloarea** [comment]
@vietj  I have updated the table

**Comment by afloarea** [comment]
@vietj I noticed you added Distro Stack Manager under 'Internal Magick'. Is this different than the stack manager under 'Devops'? Or did I made a mistake by putting it there?

**Comment by vietj** [comment]
I just meant that stack manager is something we don't really care about.

**Comment by pmlopes** [comment]
@afloarea I've just merged the `vertx-jwt` into `vertx-auth-common` as discussed in the issue. I think you are now able to work the `vertx-auth` modules without a problem!

**Comment by vietj** [comment]
@afloarea can we try getting the property in the parent pom for multi module projects when it is possible ?

**Comment by afloarea** [comment]
Thanks @pmlopes . I have raised a PR for auth.

@vietj Yes, I actually don't know why I didn't do that in the first place 😅. I have updated the auth and service discovery to use the parent pom. I will update the others as well

**Comment by afloarea** [comment]
@vietj @pmlopes I can't build the jars for vertx-web anymore, so I coudn't test with the parent pom.
It fails to compile because it can't find io.vertx.ext.jwt.*.
I also merged from upstream vertx-web but it still failed

**Comment by pmlopes** [comment]
Ah, that's my bad, there's a reference to a class from the merged modules in auth. I'll fix it...

**Comment by pmlopes** [comment]
web should be building correctly now...

**Comment by afloarea** [comment]
I tested web with parent pom and I updated the PR. Thank you @pmlopes :+1:

**Comment by vietj** [comment]
I believe I merged all the open PR @afloarea keep the PR flow coming in :-)

**Comment by vietj** [comment]
Merged all submitted PR's @afloarea

**Comment by afloarea** [comment]
@vietj I have raised PRs for most of the approved names. 

The ones missing are for kotlin. I have a question about that: should I only add for kotlin lang and coroutines? or should I wait for the kotlin.gen one as well. Also, what about the vertx-kotlin-compiler? I did not add it to the table but I noticed it in the repository.

Also, the vertx-codegen was a bit different than the others because it's using the assembly plugin so I did not use the parent pom there (FYI).

Apart from these, I believe there are no other approved names, only postponed and marked as 'under discussion' or 'Pending'.

**Comment by vietj** [comment]
for kotlin I think for now indeed we only add kotlin lang and coroutines indeed -  I merged all open PR's

**Comment by vietj** [comment]
kotlin done

**Comment by afloarea** [comment]
@vietj since all the approved names have merged PRs, what's next? I assume this issue will remain open until the postponed/under discussion names are resolved too, right?

Also, if you want, I could also have a look at merging the vertx-sql-common into vertx-jdbc-client.

**Comment by vietj** [comment]
for sql-common, I will have a look soon.

besides, I think what we need now is to capture the current state in a wiki page and write some documentation that defines the expectations and possibly at some point we need an how-to with JPMS for Vert.x 4 using automatic module names (https://how-to.vertx.io) that could be based on 4.0.0-milestone5 (cc @jponge)

**Comment by vietj** [comment]
I will create a wiki page with the current names soon I think in the vertx wiki

**Comment by vietj** [comment]
Here is the wiki page https://github.com/vert-x3/wiki/wiki/JPMS

**Comment by vietj** [comment]
I will close this issue for Vert.x 4

**Comment by pmlopes** [comment]
@vietj I've just noticed that as we missed `vertx-bridge-common` in web a jpms application will fail to build if it attempts to configure the eventbus sockjs bridge as the classes are not made public.

**Comment by pmlopes** [comment]
I've re-open it because we faced a issue with:

https://github.com/vert-x3/vertx-bridge-common/pull/6

Can we settle on a name for it?

`io.vertx.bridge.common` or `io.vertx.eventbus.bridge.common` ?

**Comment by overshareware** [comment]
Is there any information on the status of the rxjava packages? They're listed as 'postponed' but I can't find any more information on why or if that will be resolved?

**Comment by vietj** [comment]
I don't remember why we didn't do it for rxjava

**Comment by vietj** [comment]
I'd like to remember the reason and get the context

**Comment by afloarea** [comment]
@vietj Do you think it is worth adding automatic module names for the newer projects? I can raise some PRs if so.
- Web Proxy[Tech Preview] (`vertx-web-proxy`) -> `io.verx.web.proxy`
- URI Template[Tech Preview] (`vertx-uri-template`) -> `io.vertx.uri.template`
- Json Schema[Tech Preview] (`vertx-json-schema`) -> `io.vertx.json.schema`
- Http Proxy[Tech Preview] (`vertx-http-proxy`) -> `io.vertx.http.proxy`

Regarding rxjava, I unfortunately do not remember why it was postponed.

**Comment by vietj** [comment]
rather

- io.vertx.web.proxy
- io.vertx.uritemplate
- io.vertx.jsonschema
- io.vertx.httpproxy

**Comment by vietj** [comment]
@afloarea can you provide PR soon ? we will release 4.4 tomorrow

we should also have io.vertx.web.openapi.router

I added Todo to the initial list in the issue.

**Comment by vietj** [comment]
I'm closing this issue, I moved the content to the vertx wiki https://github.com/eclipse-vertx/vert.x/wiki/Automatic-Module-Names

**Comment by vietj** [comment]
Actually I've done them myself for 4.4 as it was o the critical path of the
release.
